types/stylelint/index.d.ts

```diff
@@ -252,6 +253,26 @@ declare namespace stylelint {
 		fixable?: boolean;
 	};
 
+	type FixerArguments = { [x: string]: any };
+	type Range = {
+		start?: {
+			line: number;
+			column?: number /** inclusive */;
+		};
+		end: {
+			line: number;
+			column?: number /** exclusive */;
+		};
+	};
+
+	export type Fixer = (args?: FixerArguments) => Range;
+	export type FixerData = {
+		source: { start: Required<Exclude<Range['start'], undefined>>; end: Required<Range['end']> };
+		callback: Fixer;
+		args?: FixerArguments;
+		unfixable?: boolean;
```

Comment:
[ask] Is `unfixable` necessary? If `unfixable` is `true`, we don't need to add a fixer, right?

Reply:
Note that `context.fix` does not equal `meta.fixable`.

Here, we can refer to the ESLint API:
- [How to report problems](https://eslint.org/docs/latest/extend/custom-rules#reporting-problems)
  - `context.report()` provides the `fix` function property
- [`LintMessage` type](https://eslint.org/docs/latest/integrate/nodejs-api#-lintmessage-type)
  - This is almost the same as Stylelint's `Warning` type:
  - https://github.com/stylelint/stylelint/blob/d902c0b7cc1233b74cc4959eff864cb3e2d87f73/types/stylelint/index.d.ts#L480

I believe the Stylelint API would also become similar to ESLint's one.

Reply:
> Note that `context.fix` does not equal `meta.fixable`.

I know.

> I believe the Stylelint API would also become similar to ESLint's one.

Maybe ESLint has rules that can be fixed only in certain cases like us.

> fix (EditInfo | undefined)
> The EditInfo object of autofix. **This property is undefined if this message is not fixable.**

Reply:
Still, I don't understand why we cannot use `report()` to add a fixer. 🤔 

To achieve #7192, we should add autofix info to our `Warning` type, right? For this purpose, I think the already exposed `report()` API is most appropriate instead of adding a new API like `addFixer()`, because plugin authors' learning curve may be kept lower.

Reply:
> Still, I don't understand why we cannot use `report()` to add a fixer. 🤔

You must have missed https://github.com/stylelint/stylelint/pull/7706#discussion_r1617595101.
You should review our rules, you will notice that we often have this pattern

```js
if (context.fix) {
	/* … */
	return;
}

report({
	/* … */
})
```

In short the `fixer` is never added through `report` because we are in the fix path not in the report path.
I anticipate your next question if you want to put it on `rule` it won't work either because you will lack the dynamic aspect.
i.e. the `fixer callback` must be added while the rule is being run

Reply:
Can we not change `report()` internally to support a fixer, as ESLint does?

Reply:
> Can we not change `report()` internally to support a fixer, as ESLint does?

I don't think we can without revamping how every rule currently works.
This would completely out of scope for this refactor and frankly quite the undertaking.

<hr/>

`addFixer` will exposed as an util to the plugin author here:

https://github.com/stylelint/stylelint/blob/d902c0b7cc1233b74cc4959eff864cb3e2d87f73/package.json#L43

and users may use the `--safe-disables` flag once we are done.

Reply:
I am just saying plugin authors will have a way to refactor their rules as well.

off topic

`--safe-disables` will fallback to the safe behaviour _if_ the config contains a rule which has not been transitioned else it will use the fixers in files that contain stylelint commands.
For that to work I was planning to add a `fixer: boolean` to `RuleMeta`.
The default being falsy if missing, we could easily detect plugin rules that have not yet been refactored.
Hopefully that helps you envision how everything will end up.

Reply:
I think we can choose any path here. I don't agree that it is impossible to add this to `report()`. Not saying that it must be part of `report()` persé, but rather that both options are available and we should weigh the benefits of both.

----------

How I see it :

Code you write in a Stylelint rule itself is "user land" code.

However when you call `report()` you are reporting information to Stylelint and handing over control back to the framework.

The framework then decides what to do with the report based on provided info, user preferences, disable comments, ...

From that perspective I think it makes sense to include the fixers into `report()`.

That doesn't imply that auto fixes need to use `report()`.
Only those that want to make use of this new API and have support for all the niceties that come with it need to update.

Any other rule will work as before and can still do anything they want. Rules remain "user land" code and rule authors retain the ability to mutate PostCSS AST nodes in rules.

(even today rule authors can write code that ignores `context.fix` and don't have to set `meta.fixable = true`)

Integrating into `report()` means that authors don't even need to add conditions for fixes.

Migrating to this new API could like roughly like this :

_ignore the exact field naming and API shape, the dropped `if` is more interesting than what is passed to `report()`_

```diff
- if (context.fix) {
- 	node.value = expectedFunctionName;
- 	setDeclarationValue(decl, parsed.toString());
- 
- 	return;
- }

report({
	message: messages.expected,
	messageArgs: [functionName, expectedFunctionName],
	node: decl,
	index,
	endIndex,
	result,
	ruleName,
+ 	fix: () => {
+ 		node.value = expectedFunctionName;
+ 		setDeclarationValue(decl, parsed.toString());
+ 	},
+ 	fixRange: decl.rangeBy({ index, endIndex }),
});
```

Another benefit of leveraging `report()` is that it blurs the lines between fixes en reports. Things like a dry run then become quasi trivial to implement.

--------

On the other hand I wonder if `fix()` isn't better than `addFixer()`?

This would be nicer imho:
- `report({...})`
- `fix({...})`

I do think it is important that these closely mirror each other.
Currently `addFixer` takes a list of arguments while `report` takes a single object.
They should both take a single object.

Reply:
> On the other hand I wonder if `fix()` isn't better than `addFixer()`?

I chose `addFixer` so that the temporality is clear.
i.e. I would expect `fix` to fix immediately
The notion of queuing/collection is important to convey.
Could this be considered a detail of implementation?
Maybe it's better to hide that to the rule's author.
I am not sure.

> I do think it is important that these closely mirror each other.
> Currently `addFixer` takes a list of arguments while report takes a single object.
> They should both take a single object.

You are correct; I will change its signature.

<hr />

Concerning `report`, making the changes you are describing implies

1. changing how rules are expected to work
2. fixing existing tests
3. fixing documentation
e.g. https://stylelint.io/developer-guide/rules#add-autofix, https://stylelint.io/developer-guide/plugins#the-anatomy-of-a-plugin
4. conflating reporting with fixing
5. giving `report` access to `context.fix`
6. `report` not calling `result.warn` if `context.fix` is `true`
7. increasing the diff of all refactors (longer to review)
8. etc.

I am aiming for these 33 refactors to be easy to undergo.
i.e. 2 or 3 developers should be able to tackle them in a single release
Some rules—e.g. `function-calc-no-unspaced-operator`—are already quite hard to migrate using `addFixer`.
I am trying to be pragmatic here.

**I am not saying it's not doable but that list seems daunting to me.**

`report`'s description is `Report a problem.`.
I can be interpreted in 2 ways:
 - reports a problem to stylelint
 - reports a problem to the user

What it actually does is `Creates an instance of Warning and adds it to Result#messages`.
That doesn't sound like it should have the responsibility of collecting fixes.